RNA-Seq pre-processing pipeline
===============================

Pipeline developed with Nextflow for the pre-processing
of raw FASTQ files following GATK's best practices.

Requirements
----------------

1) Nextflow can be used on any POSIX compatible system (Linus, OS X, etc.). 
It requires Bash 3.2 (or later).

2) Java 8 (or later, up to 15).

3) Windows system is supported through Windows Subsytem for Linux (WSL)
` 
    
Execution
------------------

1) Download Nextflow (version 20.01.0 or later)

    `wget -qO- https://get.nextflow.io | bash`
    
    This will create the nextflow main executable file in the current directory.

2) Make the binary executable

	`chmod +x nextflow`
	
	Optionally, move the file to a directory accessible by you path.

3) Pull the corresponding image: 

    `docker pull santiagomillan/tfg_pipeline:1.0`


4) Launch the pipeline execution: 

    `nextflow run https://github.com/santimillang/tfg_pipeline -with-docker -with-report`
    
Input/output and tools
----------------
The 'data' directory contains all the example material necessary for the execution of the workflow. 
It consists of four FASTQ files that condense the RNA expression of the experiment (paired end), as well 
as the reference genome against which they will be aligned and the annotation file used for building
the index.

The output of the pipeline consists of the index directory generated by STAR,
analysis ready BAM files, Picard MarkDuplicates' metrics and MultiQC's final report.

Tools and steps:

1) FastQC for quality control. More here: https://www.bioinformatics.babraham.ac.uk/projects/fastqc/

2) STAR for index build, mapping and sorting. More here: https://github.com/alexdobin/STAR

3) GATK's Picard suite for post-alignment processing. More here: https://broadinstitute.github.io/picard/

4) MultiQC for general report. More here: https://multiqc.info
